### PR TITLE
[JIT] add support for overloading functions

### PR DESCRIPTION
**Pull request by eellison**
This is a series of PRs that will allow us to support adding [padding to conv](https://github.com/pytorch/pytorch/pull/22484) and also reduce the friction of adding method overloads that was brought up in  https://github.com/pytorch/pytorch/pull/23266.

Support for overloaded functions following the specification in [PEP 484](https://www.python.org/dev/peps/pep-0484/#function-method-overloading).

The usage is:
```
@torch.jit.overload
def add(x: int, y: int) -> int: ...
@torch.jit.overload
def add(x: float, y: float) -> float: ...

def add:
    return x + y
```

Follow up PRs:

- Add same API for methods
- A couple of cleanups for functions:
     - don't require default params specified on the overload as well
     - potentially error if invocation could be matched to multiple overloads. now it just chooses the first one, mypy does the same thing currently
 

Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#23886 [JIT] add support for overloading functions**
* #23885 [JIT] metacompile isinstance checks

Differential Revision: [D16694863](https://our.internmc.facebook.com/intern/diff/D16694863)

### CODE REVIEW

File: torch/jit/__init__.py [hunk 1]
```diff
@@ -2079,6 +2079,80 @@ def _get_script_class(name):
...
+# qualified_name => list[(overload decl, overload default args)]
+_overloaded_fns = {}
+
+# qualified name => list[compiled fns]
+_compiled_overloaded_fns = {}
+
+def overload(func):
+    qual_name = _qualified_name(func)
+    global _overloaded_fns
+    li = _overloaded_fns.get(qual_name)
```
Comment (driazati) [inline torch/jit/__init__.py]: Can you use a more descriptive name for `li`

File: torch/jit/__init__.py [hunk 2]
```diff
@@ -2079,6 +2079,80 @@ def _get_script_class(name):
...
+    signature = torch.jit.annotations.get_signature(func)
+    if signature is None:
+        raise RuntimeError("Must explicitly add type annotations to overloaded functions: {obj}").format(func)
+    li.append((torch.jit.get_jit_def(func).decl(), get_default_args(func)))
+    return func
+
+def compile_function_with_overload(qual_name, impl_fn, overload_decl, overload_defaults):
+    _frames_up = 0
+    impl_ast = torch.jit.get_jit_def(impl_fn)
+    # TODO: figure out correct closure_rcb, stack_rcb, although this seems to work
```
Comment (driazati) [inline torch/jit/__init__.py]: You probably need to leave this as is, see the comment in `_rcb` below, but we should at least avoid copy pasting this code to make a rcb from `script`

File: torch/jit/__init__.py [hunk 3]
```diff
@@ -2079,6 +2079,80 @@ def _get_script_class(name):
...
+    # TODO: use default args from the implementation, not from the overload
+    # This is more complicated because you can have a default arg with a type
+    # incompatible with a type of parameter in an overload, and other validation.
+    # This is still an internal api so for now use defaults from overload
+    for overload_decl, overload_defaults in overloads:
+        compiled_fn = compile_function_with_overload(qual_name, obj, overload_decl, overload_defaults)
+        compiled_fns.append(compiled_fn)
+
+    # cache compilation, remove information stored to do compilation
+    _compiled_overloaded_fns[qual_name] = compiled_fns
```
Comment (driazati) [inline torch/jit/__init__.py]: This looks like a potential memory leak, if the functions this compiles go out of scope, their compiled versions will stick around here. @Chillee found a similar issue recently, I don't remember if he found a fix or not

Comment (eellison) [reply torch/jit/__init__.py]: Discussed in person: this is not a public api yet, and will have very limited usage here so we are going to leave it as is for now.

File: test/test_jit.py [hunk 4]
```diff
@@ -12832,6 +12832,91 @@ def test_non_primitive_types(x):
         out = test_non_primitive_types(_MyNamedTuple(value=torch.tensor(5.0)))
         self.assertEqual(out, torch.tensor(6.0))
 
+    def test_function_overloads(self):
+        # TODO: pyflakes currently does not compose @overload annotation with other
```
[the remark this answers is not in the document]
Comment (eellison) [reply test/test_jit.py]: Will do when this lands

File: torch/jit/__init__.py [hunk 5]
```diff
@@ -2079,6 +2085,71 @@ def _get_script_class(name):
...
+    return compiled_fns
+
+def check_directly_compile_overloaded(obj):
+    qual_name = _qualified_name(obj)
+    global _compiled_overloaded_fns
+    global _overloaded_fns
+    if qual_name in _compiled_overloaded_fns or qual_name in _overloaded_fns:
+        raise RuntimeError("Function {} cannot be directly compiled because it"
+                           " is overloaded. Please wrap it with typed inputs of"
+                           " the signature you wish to compile instead".format(obj))
```
Comment (Chillee) [inline torch/jit/__init__.py]: I think this should be `qual_name` right?

**Comment by Chillee** [comment]
Also, can you add a full example somewhere? I'm having trouble getting this to work.

I have 
```
@torch.jit.overload  # noqa: F811
def identity(x1):  # noqa: F811
    # type: (str) -> str
    pass

@torch.jit.overload  # noqa: F811
def identity(x1=1.0):  # noqa: F811
    # type: (float) -> float
    pass

def identity(x1=1.0):  # noqa: F811
    return x1

print(torch.jit.script(identity))
```

This throws an error:
```
RuntimeError: Function <function identity at 0x7fba94ef41e0> cannot be directly compiled because it is overloaded. Please wrap it with typed inputs of the signature you wish to compile instead
```

**Comment by eellison** [comment]
> Also, can you add a full example somewhere? I'm having trouble getting this to work.
> 
> I have
> 
> ```
> @torch.jit.overload  # noqa: F811
> def identity(x1):  # noqa: F811
>     # type: (str) -> str
>     pass
> 
> @torch.jit.overload  # noqa: F811
> def identity(x1=1.0):  # noqa: F811
>     # type: (float) -> float
>     pass
> 
> def identity(x1=1.0):  # noqa: F811
>     return x1
> 
> print(torch.jit.script(identity))
> ```
> 
> This throws an error:
> 
> ```
> RuntimeError: Function <function identity at 0x7fba94ef41e0> cannot be directly compiled because it is overloaded. Please wrap it with typed inputs of the signature you wish to compile instead
> ```
> 
> Specifically, it's not clear to me what "wrapping it with typed inputs" means.
> 
> EDIT: I figured it out by looking at the screenshot you initially sent me. I'd suggest rephrasing the error as something like:
> 
> ```
> RuntimeError: Function <qualname> cannot be directly compiled because it is overloaded. It must be used in the context of a function where `isinstance` checks can statically determine the type.
> ```
> 
> Perhaps less technically involved. An example might even be a good idea.

Thanks for the comments. Updated the error message, let me know if you think it's easier to understand. Btw, I suspect this is a very uncommon case (compiling a single function from nn/functional by itself).

File: test/test_jit.py [hunk 6]
```diff
@@ -12832,6 +12832,113 @@ def test_non_primitive_types(x):
...
+
+        def invoke():
+            return identity(), identity(.5), identity("hi")
+
+        self.checkScript(invoke, ())
+
+        def schema_match_failure():
+            return identity((1, 2))
+
+        thrown = False
```
Comment (driazati) [inline test/test_jit.py]: Why not `self.assertRaisesRegex`?

File: test/test_jit.py [hunk 7]
```diff
@@ -12832,6 +12832,113 @@ def test_non_primitive_types(x):
...
+            pass
+
+        @torch.jit.overload  # noqa: F811
+        def my_conv(x, y=2.0):  # noqa: F811
+            # type: (float, float) -> (float)
+            pass
+
+        def my_conv(x, y=2.0):  # noqa: F811
+            if isinstance(y, str):
+                return 4.0 - x
```
Comment (driazati) [inline test/test_jit.py]: Can you also test the `str` path here?

File: torch/csrc/jit/script/init.cpp [hunk 8]
```diff
@@ -173,6 +173,24 @@ std::shared_ptr<PythonResolver> pythonResolver(
   return std::make_shared<PythonResolver>(
       rcb, std::move(classname), std::move(classType));
 }
+
+void checkOverloadDecl(const Decl& new_decl, const Decl& old_decl) {
+  const auto& new_params = new_decl.params();
+  const auto& old_params = old_decl.params();
+
+  TORCH_INTERNAL_ASSERT(
```
Comment (driazati) [inline torch/csrc/jit/script/init.cpp]: It doesn't really matter for our use case but add a comment here that this isn't strictly necessary, it looks like the overloads can have [different signatures](https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading)

File: torch/csrc/jit/script/init.cpp [hunk 9]
```diff
@@ -702,6 +720,34 @@ void initJitScriptBindings(PyObject* module) {
...
+      "_jit_script_compile_overload",
+      [](const std::string& qualname,
+         const Decl& overload_decl,
+         const Def& implementation_def,
+         ResolutionCallback rcb,
+         const FunctionDefaults& defaults) {
+        const auto name = c10::QualifiedName(qualname);
+        auto cu = get_python_cu();
+        checkOverloadDecl(overload_decl, implementation_def.decl());
+        auto new_def = implementation_def.withDecl(overload_decl);
```
Comment (driazati) [inline torch/csrc/jit/script/init.cpp]: It looks like replacing the `implementation_def` is all this does differently from `_jit_script_compile`, that should be moved to a function so we don't have to copy paste it here

File: torch/jit/__init__.py [hunk 10]
```diff
@@ -2079,6 +2085,68 @@ def _get_script_class(name):
...
+    if fn_overload_list is None:
+        fn_overload_list = []
+        _overloaded_fns[qual_name] = fn_overload_list
+    signature = torch.jit.annotations.get_signature(func)
+    if signature is None:
+        raise RuntimeError("Must explicitly add type annotations to overloaded functions: {obj}").format(func)
+    fn_overload_list.append((torch.jit.get_jit_def(func).decl(), get_default_args(func)))
+    return func
+
+def compile_function_with_overload(qual_name, impl_fn, overload_decl, overload_defaults):
```
Comment (driazati) [inline torch/jit/__init__.py]: These would all appear in `torch.jit`, so prefix everything but `overload` with a `_`

**Review by driazati — APPROVED** [review]
Looks good, remaining stuff is mostly just structural

File: torch/csrc/jit/script/init.cpp (hunk 8 above)
Comment (eellison) [reply torch/csrc/jit/script/init.cpp]: Yup, this is along the same path as not requiring default params on the overload that is a follow up

File: test/test_jit.py (hunk 6 above)
Comment (eellison) [reply test/test_jit.py]: Couldn't figure out the regex lol

Comment (Chillee) [reply test/test_jit.py]: Would look something like this if you wanna do it with regex (gotta use lookaheads though): https://regexr.com/4irtv

`(?=.*of type 'str')(?=.*of type 'float').*`

Might not be worth it.

File: torch/csrc/jit/script/python_sugared_value.cpp [hunk 11]
```diff
@@ -544,6 +573,13 @@ std::shared_ptr<SugaredValue> toSugaredValue(
 
   py::bool_ isFunction = py::module::import("inspect").attr("isfunction")(obj);
   if (py::cast<bool>(isFunction)) {
+    auto overloads =
```
Comment (suo) [inline torch/csrc/jit/script/python_sugared_value.cpp]: Looks like the infrastructure for overloaded functions is only implemented for Python. Do we have a plan for resolving overloaded functions defined from the string frontend (for example, how will we serialize overloaded functions)?

File: torch/jit/__init__.py [hunk 12]
```diff
@@ -2110,6 +2116,68 @@ def _get_script_class(name):
...
+    _rcb = _gen_rcb(impl_fn, _frames_up)
+    fn = torch._C._jit_script_compile_overload(qual_name, overload_decl, impl_ast, _rcb, overload_defaults)
+    return fn
+
+def _get_overloads(obj):
+    # check for cached compiled fns
+    qual_name = _qualified_name(obj)
+    global _compiled_overloaded_fns
+    compiled_overloads = _compiled_overloaded_fns.get(qual_name, None)
+    if compiled_overloads is not None:
```
Comment (suo) [inline torch/jit/__init__.py]: This doesn't work if we call `_get_overloads()` then add another overload after. The second overload will not ever get compiled. Do we need a cache here? We don't have anything similar for other recursively compiled functions.

File: torch/csrc/jit/script/init.cpp [hunk 13]
```diff
@@ -423,8 +463,8 @@ void initJitScriptBindings(PyObject* module) {
       .def(
           "_register_attribute",
           [](Module& self, std::string name, TypePtr type, py::object value) {
-            auto unshaped = unshapedType(type);
-            self.register_attribute(name, unshaped, toIValue(value, type));
+            self.register_attribute(
```
Comment (suo) [inline torch/csrc/jit/script/init.cpp]: Why was this changed?

Comment (eellison) [reply torch/csrc/jit/script/init.cpp]: Clang tidy complained

File: torch/jit/__init__.py (hunk 12 above)
Comment (eellison) [reply torch/jit/__init__.py]: What is the use case you are imagining? This is still an internal only api so I don't think it applies but something to think about if we make it public. We don't strictly need a cache it's true.

File: torch/csrc/jit/script/python_sugared_value.cpp (hunk 11 above)
Comment (eellison) [reply torch/csrc/jit/script/python_sugared_value.cpp]: Any overloaded function that needs to be serialized will already be instantiated in one of its overloads, that function gets serialized normally.

File: torch/jit/__init__.py (hunk 12 above)
Comment (suo) [reply torch/jit/__init__.py]: If we add a new overload after using the overload in the scripted function, it will not be registered. Let's remove the cache, or make the existing caching logic correct. Even if it's unlikely I think it'll be very surprising if we ever run into this.
